binsync/configuration.py

```diff
@@ -1,164 +1,94 @@
 
 l = logging.getLogger(__name__)
 
 
-class Config:
+class ProjectData:
     __slots__ = (
-        "path",
+        "binary_name",
+        "user",
+        "repo_path",
+        "remote",
     )
 
-    def __init__(self, path):
-        self.path = path
-
-    def save(self):
-        self.path = pathlib.Path(self.path).expanduser().absolute()
-        if not self.path.parent.exists():
-            return None
-
-        dump_dict = {}
-        for attr in self.__slots__:
-            attr_val = getattr(self, attr)
-            if isinstance(attr_val, pathlib.Path):
-                attr_val = str(attr_val)
-
-            dump_dict[attr] = attr_val
-
-        with open(self.path, "w") as fp:
-            toml.dump(dump_dict, fp)
-
-        return self.path
-
-    def load(self):
-        self.path = pathlib.Path(self.path).expanduser().absolute()
-        if not self.path.exists():
-            return None
-
-        with open(self.path, "r") as fp:
-            load_dict = toml.load(fp)
-
-        for attr in self.__slots__:
-            setattr(self, attr, load_dict.get(attr, None))
-
-        return self
-
-    @classmethod
-    def load_from_file(cls, path):
-        conf = cls(path)
-        return conf.load()
-
-    @classmethod
-    def update_or_make(cls, path, **attrs_to_update):
-        path = pathlib.Path(cls.correct_path(path)).expanduser().absolute()
-        # find or create a new config
-        conf = cls.load_from_file(path) if path.exists() \
-            else cls(path)
+    def __init__(self, binary_name, user=None, repo_path=None, remote=None):
+        self.binary_name = binary_name
+        self.user = user
+        self.repo_path = repo_path
+        self.remote = remote
 
-        # update every value in the Config
-        for attr, val in attrs_to_update.items():
-            if attr in conf.__slots__:
-                setattr(conf, attr, val)
+    def __setstate__(self, data):
+        for k in self.__slots__:
+            if k in data:
+                setattr(self, k, data[k])
 
-        conf.save()
-        return conf
+    def __getstate__(self):
+        return dict(
+            (k, getattr(self, k)) for k in self.__slots__
+        )
 
     @classmethod
-    def correct_path(cls, path):
-        return path
+    def get_from_state(cls, data):
+        proj_data = cls(data['binary_name'])
+        proj_data.__setstate__(data)
+        return proj_data
 
 
-class ProjectConfig(Config):
-    __slots__ = Config.__slots__ + (
-        "binary_name",
-        "user",
-        "repo_path",
-        "remote",
+class BinSyncBSConfig(BSConfig):
+    __slots__ = BSConfig.__slots__ + (
+        "recent_projects",
         "table_coloring_window",
         "log_level",
-        "merge_level"
+        "merge_level",
     )
 
     def __init__(self,
-                 binary_path,
-                 user=None,
-                 repo_path=None,
-                 remote=None,
+                 save_location=None,
```

Comment:
If possible, type annotate things that are arguments to a class or function. In this case, `save_location` could be annotated like so:
```py
save_location: Optional[Path] = None
```